azure-iot-hub-devicesdk/samples/send_telemetry_device.py

```diff
@@ -3,48 +3,59 @@
+    from_connection_string,
+)  # this is a overlong import, fix
 
-# The connection string for a device should never be stored in code. For the sake of simplicity we're using an environment variable here.
-conn_str = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
-# The "Authentication Provider" is the object in charge of creating authentication "tokens" for the device client.
-auth_provider = from_connection_string(conn_str)
-# For now, the SDK only supports MQTT as a protocol. the client object is used to interact with your Azure IoT hub.
-# It needs an Authentication Provider to secure the communication with the hub, using either tokens or x509 certificates
-device_client = DeviceClient.from_authentication_provider(auth_provider, "mqtt")
+logging.basicConfig(level=logging.INFO)
 
+messages_to_send = 10
 
-# The connection state callback allows us to detect when the client is connected and disconnected:
-def connection_state_callback(status):
-    print("connection status: " + status)
 
+async def main():
+    # The connection string for a device should never be stored in code. For the sake of simplicity we're using an environment variable here.
+    conn_str = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
 
-# Register the connection state callback with the client...
-device_client.on_connection_state = connection_state_callback
+    # The "Authentication Provider" is the object in charge of creating authentication "tokens" for the device client.
+    # TODO: open question: do we want async versions of from_connection_string and from_authentication_provider?
+    auth_provider = from_connection_string(conn_str)
 
-# ... and connect the client.
-device_client.connect()
+    # For now, the SDK only supports MQTT as a protocol. the client object is used to interact with your Azure IoT hub.
+    # It needs an Authentication Provider to secure the communication with the hub, using either tokens or x509 certificates
+    device_client = await DeviceClient.from_authentication_provider(auth_provider, "mqtt")
 
-# send 5 messages with a 1 second pause between each message
-for i in range(0, 5):
-    print("sending message #" + str(i))
-    msg = Message("test wind speed " + str(i))
-    msg.message_id = uuid.uuid4()
-    msg.correlation_id = "correlation-1234"
-    msg.custom_properties["tornado-warning"] = "yes"
-    device_client.send_event(msg)
-    time.sleep(1)
+    # The connection state callback allows us to detect when the client is connected and disconnected:
+    def connection_state_callback(status):
+        print("connection status: " + status)
 
-# send only string messages
-for i in range(5, 10):
-    print("sending message #" + str(i))
-    device_client.send_event("test payload message " + str(i))
-    time.sleep(1)
+    # Register the connection state callback with the client...
+    # TODO: open question: connection_state_callback is a sync callback.  Do we want to support async callbacks?
```

Comment:
># TODO: open question: connection_state_callback is a sync callback.  Do we want to support async callbacks? [](start = 4, length = 108)

Lets close on both the TODOs in this file ASAP.  Is there any reason to not do async callbacks and async factory methods?  I think async callbacks are required (otherwise customers need to jump through hoops to call async code) but I'm only 90% convinced about the factory methods.

Reply:
This may require some teamwork to address - I largely left this functionality alone in the overhaul, because I wasn't entirely clear on the implications of making changes related to callbacks. I think this will require a separate PR, however I agree that it 100% should be addressed.

---
In reply to: [246911794](https://github.com/Azure/azure-iot-sdk-python-preview/pull/39#discussion_r246911794) [](ancestors = 246911794)